### PR TITLE
68 feature implement traffic light control

### PR DESCRIPTION
**Pull request by MarcoRiedenauer**
# Description

This PR adds the implementation of the traffic light detection. It enables to classify traffic lights into the states green, yellow, red, backside.
Unfortunately it is not possible for me to implement DVC till tomorrow, so there are "some" images added to git ^^

## Time invested

Marco additional time: 1.5h

## Type of change

Please delete options that are not relevant.

- New feature (non-breaking change which adds functionality)

## Does this PR introduce a breaking change?

No, since the traffic light classification is not used yet.

## Most important changes

Most important changes are the traffic_light_training.py and classification_model.py

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works (might be obsolete with CI later on)
- [x] New and existing unit tests pass locally with my changes (might be obsolete with CI later on)

### CODE REVIEW

File: code/perception/src/traffic_light_detection/src/traffic_light_detection/traffic_light_training.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+        self.weights_organizer = WeightsOrganizer(cfg=self.cfg,
+                                                  model=self.model)
+
+    def run(self):
+        """
+        Trains the model for a given amount of epochs
+        """
+        self.model.to(self.cfg.DEVICE)
+        tepoch = tqdm(range(self.cfg.EPOCHS))
+        for i in range(self.cfg.EPOCHS):
```
Comment (nylser) [inline code/perception/src/traffic_light_detection/src/traffic_light_detection/traffic_light_training.py:76]: would it make sense to use https://pytorch.org/docs/1.3.0/optim.html#torch.optim.lr_scheduler.ExponentialLR instead of your own implementation?

**Review by nylser — CHANGES_REQUESTED** [review]
Thank you for your very good work, nice and clean implemenation of the traffic light model.

I think one thing that would be very helpful for reference is to have an overview of how we can use this model. 
Like a simple example of loading a traffic light image into it and getting a label back?
Or do you think it is already clear from looking at the validation code?

File: code/perception/src/traffic_light_detection/src/traffic_light_detection/traffic_light_training.py (hunk 1 above)
Comment (MarcoRiedenauer) [reply code/perception/src/traffic_light_detection/src/traffic_light_detection/traffic_light_training.py:76]: Oh yes, forgot it 🙈

**Comment by MarcoRiedenauer** [comment]
> Thank you for your very good work, nice and clean implemenation of the traffic light model.
> 
> I think one thing that would be very helpful for reference is to have an overview of how we can use this model. Like a simple example of loading a traffic light image into it and getting a label back? Or do you think it is already clear from looking at the validation code?

You’re right, that’s a good idea. It should be clear for us from the validation code, but it is maybe hard to find for others. I’ll implement a inference class for loading a model and images and classify them 👍

File: code/perception/src/traffic_light_detection/src/traffic_light_detection/traffic_light_training.py (hunk 1 above)
Comment (MarcoRiedenauer) [reply code/perception/src/traffic_light_detection/src/traffic_light_detection/traffic_light_training.py:78]: Resolved in [cb23b91](https://github.com/ll7/paf22/pull/116/commits/cb23b91f07e462d9e67e01d33f34431376feedfe)

**Review by nylser — APPROVED** [review]
Thank you for your additions!

**Comment by nylser** [comment]
PS: Don't forget to track your additional time on this PR ;)

**Comment by MarcoRiedenauer** [comment]
> PS: Don't forget to track your additional time on this PR ;)

Thanks :D Done :)
